pkg/apis/kudo/v1beta1/instance_admission.go

```diff
@@ -0,0 +1,239 @@
+| x                |                 |         |              | x          | No    | Forbid plan cancellations (for now)                                                                                 |
+| ---              |                 |         |              |            |       | ---                                                                                                                 |
+|                  | x               |         | x            |            | No    | Forbid simultaneous parameter update and directly triggered plan                                                    |
+|                  |                 | x       | x            |            | No    | Forbid simultaneous upgrades and directly triggered plans                                                           |
+|                  | x               | x       |              |            | No*   | Forbid simultaneous upgrades and parameter updates                                                                  |
+ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+ *Note: simultaneous upgrade and parameter update is NOT allowed. However, there is a exception where new OV removes an existing
+ parameter. Removing this parameter in the instance update would not count as parameter update (since there is no plan to trigger).
+
+ For the complete set of rules, see the corresponding test.
+*/
+
+// validateUpdate takes in the old and new (updated) instance and returns a new plan that might
+// be triggered based on the update and an error if the update is not valid. Return plan might be
+// - <nil> when there is no change to an existing scheduled plan
+// - '' empty string when an existing plan should be canceled (not implemented yet)
+// - 'newPlan' some new plan that should be triggered
+func validateUpdate(old, new *Instance, ov *OperatorVersion) (*string, error) {
+	// PREREQUISITES:
+	newPlan := new.Spec.PlanExecution.PlanName
+	oldPlan := old.Spec.PlanExecution.PlanName
+	newOvRef := new.Spec.OperatorVersion
+	oldOvRef := old.Spec.OperatorVersion
+
+	paramDiff := ParameterDiff(old.Spec.Parameters, new.Spec.Parameters)
+	paramDefs := GetParamDefinitions(paramDiff, ov)
+	triggeredPlan, err := triggeredPlan(paramDefs, ov)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update Instance %s/%s: %v", old.Namespace, old.Name, err)
+	}
+
+	hasPlan := oldPlan != ""
+	isParameterUpdate := triggeredPlan != nil
+	isUpgrade := newOvRef != oldOvRef
+	isNovelPlan := !hasPlan && newPlan != ""
+	isPlanOverride := hasPlan && newPlan != "" && newPlan != oldPlan
+	isPlanCancellation := hasPlan && newPlan == ""
+
+	// Checking compatibility rules described in the table above:
+	switch {
+	case hasPlan && isParameterUpdate && *triggeredPlan != oldPlan:
```

Comment:
isn't `hasPlan && isParameterUpdate` enough? If there's plan running, we don't want to allow update in params even if it's the same plan, right? (that would mean that manifests change mid-execution)

Reply:
hmm well the question is - when is the PlanExecution cleared? Is it when the plan starts or ends? I know we discussed this but I don't know what is the current thinking and it is important for these rules...

EDIT: Ok I see that the docs for the field say that it will be cleared when execution finishes so my previous comment is I think relevant

Reply:
EDIT2: Oh I see ... https://github.com/kudobuilder/kudo/pull/1188/files#diff-3ee7e3fb450b5a364e3d9139fb154ad2R155 ... I need to think about that :)

Reply:
>  If there's plan running, we don't want to allow update in params even if it's the same plan, right?

I believe, we do. This is a scenario when e.g. a deployment is stuck because too many instances are set (or too much mem). You would want to update a param and "re-trigger" the plan.